### PR TITLE
CR#23065 Fix revision handling for Teamscale upload

### PR DESCRIPTION
**Pull request by DreierF**
Addresses issue #

- [ ] Changes are tested adequately
- [ ] Agent's README.md updated in case of user-visible changes
- [ ] CHANGELOG.md updated

Please respect the vote of the [Teamscale bot](https://demo.teamscale.com) or flag irrelevant findings as tolerated or false positives. If you feel that the Teamscale config needs adjustment, please state so in a comment and discuss this with your reviewer.

### CODE REVIEW

File: agent/src/main/java/com/teamscale/jacoco/agent/testimpact/TestEventHandlerStrategyBase.java [hunk 1]
```diff
@@ -90,7 +90,11 @@ public String testRunStart(List<ClusteredTestDetails> availableTests, boolean in
 					" Thus, you cannot use the agent to retrieve impacted tests via the testrun/start REST endpoint." +
 					" Please use the 'teamscale-' agent parameters to configure a Teamscale connection.");
 		}
-
+		if (agentOptions.getTeamscaleServerOptions().commit == null) {
+			throw new UnsupportedOperationException("You did not provide a 'teamscale-commit' or " +
+					"'teamscale-commit-manifest-jar'. 'teamscale-revision' is not sufficient to retrieve impacted " +
```
Comment (sewe) [inline agent/src/main/java/com/teamscale/jacoco/agent/testimpact/TestEventHandlerStrategyBase.java:95]: I found it a bit odd that `AgentOptions` doesn't contain string constants for the various options, given that they are included in a lot of error messages (which is very helpful!).

File: teamscale-client/src/main/java/com/teamscale/client/TeamscaleServer.java [hunk 2]
```diff
@@ -54,9 +54,17 @@ public boolean hasAllRequiredFieldsNull() {
...
+	public boolean hasCommitOrRevision() {
+		return commit != null || revision != null;
+	}
+
 	@Override
 	public String toString() {
-		return "Teamscale " + url + " as user " + userName + " for " + project + " to " + partition + " at commit "
-				+ commit;
+		String at = "revision " + revision;
+		if (commit != null) {
```
[the remark this answers is not in the document]
Comment (DreierF) [reply teamscale-client/src/main/java/com/teamscale/client/TeamscaleServer.java:65]: This should not happen as the validation checks for this case, but I still inverted the logic

```
validator.isTrue(teamscaleServer.revision == null || teamscaleServer.commit == null,
				"'teamscale-revision' is incompatible with 'teamscale-commit' and 'teamscale-commit-manifest-jar'.");
```

File: teamscale-client/src/main/java/com/teamscale/client/TeamscaleServer.java [hunk 3]
```diff
@@ -54,9 +54,17 @@ public boolean hasAllRequiredFieldsNull() {
...
+			at = "commit " + commit;
```
[the remark this answers is not in the document]
Comment (DreierF) [reply teamscale-client/src/main/java/com/teamscale/client/TeamscaleServer.java:66]: Teamscale does not like ternary operators...

File: agent/src/main/java/com/teamscale/jacoco/agent/testimpact/CoverageToTeamscaleStrategy.java [hunk 4]
```diff
@@ -44,8 +44,8 @@ public CoverageToTeamscaleStrategy(JacocoRuntimeController controller, AgentOpti
 		super(agentOptions, controller);
 		this.reportGenerator = reportGenerator;
 
-		if (agentOptions.getTeamscaleServerOptions().commit == null) {
-			throw new UnsupportedOperationException("You must provide a commit via the agent's options." +
+		if (!agentOptions.getTeamscaleServerOptions().hasCommitOrRevision()) {
+			throw new UnsupportedOperationException("You must provide a commit or revision via the agent's options." +
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/testimpact/CoverageToTeamscaleStrategy.java:48]: Done

File: agent/src/main/java/com/teamscale/jacoco/agent/testimpact/TestEventHandlerStrategyBase.java (hunk 1 above)
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/testimpact/TestEventHandlerStrategyBase.java:95]: Started adding constants for it.

File: teamscale-client/src/main/java/com/teamscale/client/TeamscaleServer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sewe) [reply teamscale-client/src/main/java/com/teamscale/client/TeamscaleServer.java:68]: I thought we have the "trivial" ternary operator exception now, but maybe not in our analysis profile. But nevermind. The above is fine.

**Pull request merged**